clippy_lints/src/functions/mod.rs

```diff
@@ -399,6 +400,50 @@ declare_clippy_lint! {
     "renamed function parameters in trait implementation"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Warns when a function signature uses `&Option<T>` instead of `Option<&T>`.
+    ///
+    /// ### Why is this bad?
+    /// More flexibility, better memory optimization, and more idiomatic Rust code.
+    ///
+    /// `&Option<T>` in a function signature breaks encapsulation because the caller must own T
+    /// and move it into an Option to call with it. When returned, the owner must internally store
+    /// it as `Option<T>` in order to return it.
+    /// At a lower level `&Option<T>` points to memory that has `presence` bit flag + value,
+    /// whereas `Option<&T>` is always optimized to a single pointer.
```

Comment:
Remark: it would be good if we can reference any guarantees that these optimizations are performed. We should double-check `Option<T>` layout guarantees by std docs: https://doc.rust-lang.org/std/option/index.html#representation.

Reply:
thx, i updated the text with the link